### ISSUE TITLE
InputNumber: onChange parameter type

### ISSUE THREAD

**Issue by AryamnovEugeniy** [mantinedev/mantine#5037]
### What package has an issue

@mantine/core

### Describe the bug

I noticed that in the latest version onChange in InputNumber is called with a string as its parameter when defaultValue is undefined. I am not sure this is correct, and it is definitely not correct for my situation.
Although of course there is a simple workaround. The problem for me is that I do not think that current behaviour is correct.

### What version of @mantine/* packages do you have in package.json? (Note that all @mantine/* packages must have the same version in order to work correctly)

7.1.3

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/mantine-forked-xq74sl?file=/src/App.tsx

### Do you know how to fix the issue

None

### Are you willing to participate in fixing this issue and create a pull request with the fix

None

### Possible fix

I found a PR that introduced this change: https://github.com/mantinedev/mantine/pull/4916
Possibly, when [_value](https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/components/NumberInput/NumberInput.tsx#L231) is undefined, we should pass a number to onChange. What do you think? Although I did not test this solution for these issues: #4901 , #4912 
And to be honest, I am not sure that we should ever pass a string as a parameter to onChange (except maybe when there is no value, but in this case null is a better option from my point of view) - this is **Number**Input after all =)

**Comment by 7iomka** [comment]
I agree with you that the type of string inside NumberInput make weird typings and mappings.
For example in my use-cases I need to refactor onChange prop on all NumberInput based components to something like
this:
- `v?: number`
```ts
      onChange={(val) => onChange(val === '' ? undefined : (val as number))}
```

- `v: number | null`
```ts
      onChange={(val) => onChange(val === '' ? null : (val as number))}
```

- `v: number`
```ts
      onChange={(val) => onChange(val as number)}
```

The last one has issues of course.
So, I don't know what is the best way to handle empty NumberInput value with better types alongside with safety.

**Comment by odbayar** [comment]
Also the type of the value changes to **number** when you use the arrow buttons to increase and decrease the value, and it changes back to **string** when you use your keyboard to type numbers.

**Comment by 7iomka** [comment]
> Also the type of the value changes to **number** when you use the arrow buttons to increase and decrease the value, and it changes back to **string** when you use your keyboard to type numbers.

Same issue.
I think it is related with wrong or inconsistent handling of value, because react-input-format give us string and float value, but mantine not.
![telegram-cloud-photo-size-2-5334748595003446414-y](https://github.com/mantinedev/mantine/assets/18473137/770815f8-21ec-4060-b1e7-fdc3ab93ca04)

**Comment by rtivital** [comment]
Fixed in [7.2.2](https://github.com/mantinedev/mantine/releases/tag/7.2.2)
